storage/src/adb/keyless.rs

```diff
@@ -1324,4 +1367,97 @@ mod test {
             db.destroy().await.unwrap();
         });
     }
+
+    #[test_traced("WARN")]
+    fn test_keyless_db_replay_with_trailing_appends() {
+        let executor = deterministic::Runner::default();
+        executor.start(|context| async move {
+            let mut hasher = Standard::<Sha256>::new();
+
+            // Create initial database with committed data
+            let mut db = open_db(context.clone()).await;
+
+            // Add some initial operations and commit
+            for i in 0..10 {
+                let v = vec![i as u8; 10];
+                db.append(v).await.unwrap();
+            }
+            db.commit(None).await.unwrap();
+            let committed_root = db.root(&mut hasher);
+            let committed_size = db.op_count();
+
+            // Add exactly one more append (uncommitted)
+            let uncommitted_value = vec![99u8; 20];
+            db.append(uncommitted_value.clone()).await.unwrap();
+
+            // Sync only the log (not MMR or locations)
+            db.simulate_failure(true, false, false).await.unwrap();
+
+            // Reopen database
+            let mut db = open_db(context.clone()).await;
+
+            // Verify correct recovery
+            assert_eq!(
+                db.op_count(),
+                committed_size,
+                "Should rewind to last commit"
+            );
+            assert_eq!(
+                db.root(&mut hasher),
+                committed_root,
+                "Root should match last commit"
+            );
+            assert_eq!(
+                db.last_commit_loc(),
+                Some(committed_size - 1),
+                "Last commit location should be correct"
+            );
+
+            // Verify the uncommitted append was properly discarded
+            // We should be able to append new data without issues
+            let new_value = vec![77u8; 15];
+            let loc = db.append(new_value.clone()).await.unwrap();
+            assert_eq!(
+                loc, committed_size,
+                "New append should get the expected location"
+            );
+
+            // Verify we can read the new value
+            assert_eq!(db.get(loc).await.unwrap(), Some(new_value));
+
+            // Test with multiple trailing appends to ensure robustness
+            db.commit(None).await.unwrap();
+            let new_committed_root = db.root(&mut hasher);
+            let new_committed_size = db.op_count();
+
+            // Add multiple uncommitted appends
+            for i in 0..5 {
+                let v = vec![(200 + i) as u8; 10];
+                db.append(v).await.unwrap();
+            }
+
+            // Simulate the same partial failure scenario
```

Comment:
This (multiple uncommitted appends) should be well-tested by the existing recovery tests I think?  I guess it doesn't hurt to have though.

Reply:
I was thinking about modifying my test to append random # of items that get rolled back, can do that later though.

Reply:
Seems like something that may be better implemented via the `fuzzer`?

Reply:
yeah that's true